### PR TITLE
Added an optimizer and fitter 

### PR DESCRIPTION
**Pull request by nocturnalastro**
I added an optimizer for scipy's minimize optimizer and the bare bones of a fitter still needs to check and then pass jac's if required which it doesn't do but scipy will throw a warning it's required so the user could do it them self 
I also modified the fitter class to allow the passing of keywords to the optimizer on creation
I as tried to make pep8 happy

### CODE REVIEW

**Comment by pllim** [comment]
You could have just use the same #4696 PR, but never mind now.

From a glance, some things are missing:

* Change log
* Tests
* Doc string missing periods, proper capitalization, etc. Some wrong formatting.

**Comment by nocturnalastro** [comment]
I'm not sure what happened with the failed travis-cl test the log says
`/home/travis/build.sh: line 45: warning: setlocale: LC_ALL: cannot change locale (C.ascii): No such file or directory`
and then at the end 
`locale.setlocale(locale.LC_ALL, '')
File "/home/travis/miniconda/envs/test/lib/python2.7/locale.py", line 579, in setlocale
return _setlocale(category, locale)
locale.Error: unsupported locale setting`

**Comment by taldcroft** [comment]
@nocturnalastro - not your fault.  Something else is wrong and this problem is showing up all over.

File: astropy/modeling/fitting.py [hunk 1]
```diff
@@ -840,6 +838,61 @@ def _convert_input(x, y, z=None, n_models=1, model_set_axis=0):
...
+
+    Parameters
+    ----------
+    method : str or callable(custom)
+        select minimization method
+
+    supported_constraints : list
+        lists the constraint types supported by the Optimization method default is fixed
+        and tied parameters, these can be set for custom methods for predifined methods
+        this will be changed atomatically
```
Comment (hamogu) [inline astropy/modeling/fitting.py]: automatically

File: astropy/modeling/fitting.py [hunk 2]
```diff
@@ -840,6 +838,61 @@ def _convert_input(x, y, z=None, n_models=1, model_set_axis=0):
...
+    A fitter for the minimize optimizer.
+
+    Parameters
+    ----------
+    method : str or callable(custom)
+        select minimization method
+
+    supported_constraints : list
+        lists the constraint types supported by the Optimization method default is fixed
+        and tied parameters, these can be set for custom methods for predifined methods
```
Comment (hamogu) [inline astropy/modeling/fitting.py]: defined

File: astropy/modeling/fitting.py [hunk 3]
```diff
@@ -840,6 +838,61 @@ def _convert_input(x, y, z=None, n_models=1, model_set_axis=0):
...
+
+    A fitter for the minimize optimizer.
+
+    Parameters
+    ----------
+    method : str or callable(custom)
+        select minimization method
+
+    supported_constraints : list
+        lists the constraint types supported by the Optimization method default is fixed
```
Comment (hamogu) [inline astropy/modeling/fitting.py]: Just read over this again. There is also a "." missing after "method" I believe.

File: astropy/modeling/optimizers.py [hunk 4]
```diff
@@ -239,3 +241,122 @@ def __call__(self, objfunc, initval, fargs, **kwargs):
                           "Maximum number of iterations reached.",
                           AstropyUserWarning)
         return fitparams, self.fit_info
+
+
+class Minimize(Optimization):
+    """
+    Interface for Scipy.Optimize.minimize.
```
Comment (hamogu) [inline astropy/modeling/optimizers.py]: If you write
```
`scipy.optimize.minimize`
```
then sphinx will automatically generate a link when it builds the documentation. That's useful because a user might want to look up the parameters he/she can pass.

File: astropy/modeling/tests/test_fitters.py [hunk 5]
```diff
@@ -360,19 +360,46 @@ def test_param_cov(self):
...
+        fslsqp = SLSQPLSQFitter()
+        with ignore_non_integer_warning():
+            slsqp_model = fslsqp(g1, self.xdata, self.ydata)
+        model = fitter(g1, self.xdata, self.ydata)
+        assert_allclose(model.parameters, slsqp_model.parameters,
+                        rtol=10 ** (-4))
+
+    def test_minimize(self):
+        """
+        Runs MinimizeFitter with most methords.
```
Comment (mirca) [inline astropy/modeling/tests/test_fitters.py]: methods

File: astropy/modeling/fitting.py [hunk 6]
```diff
@@ -840,6 +838,61 @@ def _convert_input(x, y, z=None, n_models=1, model_set_axis=0):
...
+        model_copy = _validate_model(model, self._opt_method.supported_constraints)
+
+        farg = _convert_input(x, y, z)
+        farg = (model_copy, weights, ) + farg
+
+        p0, _ = _model_to_fit_params(model_copy)
+
+        fitparams, self.fit_info = self._opt_method(self.objective_function, p0, farg, **kwargs)
+        _fitter_to_model_params(model_copy, fitparams)
+        return model_copy
```
Comment (mirca) [inline astropy/modeling/fitting.py]: It would be nice not only to return `model_copy`, but also the `scipy.optimize.OptimizeResult` object, so that the user can get the uncertainties on the estimated parameters for free through the `hess_inv` attribute.

Comment (nocturnalastro) [inline astropy/modeling/fitting.py]: The optimizer currently returns the info to be consistent with the other astropy optimizers but it can easily return `scipy.optimize.OptimizeResult ` just a case of not filtering `res` in the return statement

File: astropy/modeling/fitting.py [hunk 7]
```diff
@@ -403,9 +402,9 @@ def objective_function(self, fps, *args):
         _fitter_to_model_params(model, fps)
         meas = args[-1]
         if weights is None:
-            return np.ravel(model(*args[2 : -1]) - meas)
+            return np.ravel(model(*args[2: -1]) - meas)
```
Comment (taldcroft) [inline astropy/modeling/fitting.py]: Even though existing PEP8 errors are annoying it is generally preferred to leave them alone if they are unrelated to your specific patch.  Then you can do a separate PR that is devoted strictly to PEP8 / style changes.  The goal is that reviewers see only changes related to the relevant patch.

Comment (nocturnalastro) [inline astropy/modeling/fitting.py]: Oh sorry I'll keep that in mind :).

**Comment by nden** [comment]
This needs rebasing. @nocturnalastro Are you interested in finishing this?

**Comment by nocturnalastro** [comment]
Yes I actually made the OptimizeResult  available recently, I'll hopefully rebase this weekend.

**Comment by nden** [comment]
Awesome!

**Comment by saimn** [comment]
It would be great to have this new fitter, any chance that you can rebase this and fix conflicts @nocturnalastro ?

**Comment by saimn** [comment]
@nocturnalastro - It looks like your rebase went bad, you probably merged with master whereas you should have force push after the rebase. You can still remove the merge commit and force push.

**Comment by nocturnalastro** [comment]
I'll give that a go :)

File: astropy/modeling/fitting.py [hunk 8]
```diff
@@ -1156,6 +1154,62 @@ def _convert_input(x, y, z=None, n_models=1, model_set_axis=0):
...
 
+class MinimizeFitter(Fitter):
+    """
+
+    A fitter for the minimize optimizer.
+
+    Parameters
+    ----------
+    method : str or callable(custom)
+        select minimization method
```
Comment (saimn) [inline astropy/modeling/fitting.py]: Could you make it more explicit what methods are supported ? Maybe just refer to the scipy documentation.

**Comment by saimn** [comment]
@nocturnalastro - you will need to rebase again to get the CI fixes, sorry (though it should be easier this time).

File: astropy/modeling/optimizers.py [hunk 9]
```diff
@@ -240,3 +242,125 @@ def __call__(self, objfunc, initval, fargs, **kwargs):
...
+            if method in ['nelder-mead', 'powell', 'cg', 'bfgs', 'newton-cg', 'dogleg', 'trust-ncg']:
+                self.supported_constraints = ['fixed', 'tied']
+            elif method in ['l-bfgs-b', 'tnc']:
+                self.supported_constraints = ['fixed', 'tied', 'bounds']
+            elif method == 'cobyla':
+                self.supported_constraints = ['fixed', 'tied', 'ineqcons']
+            elif method in ['slsqp']:
+                self.supported_constraints = ['fixed', 'tied', 'eqcons', 'ineqcons', 'bounds']
+
+            if ['newton-cg', 'trust-ncg']:
```
Comment (mirca) [inline astropy/modeling/optimizers.py]: A few more trust-region algorithms were added in `scipy` recently; maybe you want to include them as well?

File: astropy/modeling/tests/test_fitters.py [hunk 10]
```diff
@@ -501,6 +501,21 @@ def test_param_cov(self):
...
+    def test_minimize(self):
+        """
+        Runs MinimizeFitter with most methods.
+        """
+        # get rid of tiny bound so the method which don't support
+        # bounds can be tested
+        self.gauss.bounds['stddev'] = (None, None)
+        for meth in ['Nelder-Mead', 'Powell', 'CG', 'BFGS', 'L-BFGS-B', 'TNC', 'SLSQP']:
+            mm = MinimizeFitter(meth)
+            res=mm(self.gauss, self.xdata, self.ydata)
```
Comment (mirca) [inline astropy/modeling/tests/test_fitters.py]: I think you want to change the initial model a little. Right now, this test doesn't do much since it compares whether the result of the optimizer is close to the initial guess.

Comment (mirca) [reply astropy/modeling/tests/test_fitters.py]: Maybe you also want to assert that `res['status']` is meaningful?

**Comment by mirca** [comment]
This looks great to me and will be a great addition to `astropy.modeling`. I just would like to see some entries in the narrative docs and a little improvement on the unit tests.

File: astropy/modeling/optimizers.py (hunk 9 above)
Comment (saimn) [reply astropy/modeling/optimizers.py]: Also I guess this line should be `if method in ...` ? Currently it will always be True ;).

File: astropy/modeling/fitting.py [hunk 11]
```diff
@@ -1156,6 +1154,62 @@ def _convert_input(x, y, z=None, n_models=1, model_set_axis=0):
     return farg
 
 
+class MinimizeFitter(Fitter):
+    """
+
+    A fitter for the minimize optimizer.
```
Comment (nden) [inline astropy/modeling/fitting.py]: Could you make it clear this is scipy's minimize optimizer.
A reference to the documentation would be good too.

File: astropy/modeling/optimizers.py [hunk 12]
```diff
@@ -240,3 +242,125 @@ def __call__(self, objfunc, initval, fargs, **kwargs):
...
+
+    def __init__(self, method='slsqp', supported_constraints=None):
+        from scipy.optimize import minimize
+        super(Minimize, self).__init__(minimize)
+        method = method.lower()
+        if callable(method):
+            if supported_constraints is None:
+                warnings.warn("No supported_constraints set, by default the optimizer will only support fixed and tied parameters."
+                              "If you wish this optimizer to support more then you change this", AstropyUserWarning)
+
```
Comment (nden) [inline astropy/modeling/optimizers.py]: This line is too long.

File: astropy/modeling/optimizers.py [hunk 13]
```diff
@@ -240,3 +242,125 @@ def __call__(self, objfunc, initval, fargs, **kwargs):
...
+        if 'bounds' in self.supported_constraints:
+            bounds = [par.bounds for par in pars if (par.fixed is not True and par.tied is False)]
+            bounds = np.asarray(bounds)
+            for i in bounds:
+                if i[0] is None:
+                    i[0] = DEFAULT_BOUNDS[0]
+                if i[1] is None:
+                    i[1] = DEFAULT_BOUNDS[1]
+                # older versions of scipy require this array to be float
+                kwargs['bounds'] = np.asarray(bounds, dtype=np.float)
```
Comment (nden) [inline astropy/modeling/optimizers.py]: This does not need to be in the loop.

File: astropy/modeling/optimizers.py [hunk 14]
```diff
@@ -240,3 +242,125 @@ def __call__(self, objfunc, initval, fargs, **kwargs):
...
+        if res['status'] == 1:
+            warnings.warn("The fit may be unsuccessful; "
+                          "Maximum number of function evaluations reached.",
+                          AstropyUserWarning)
+        if res['status'] == 2:
+            warnings.warn("The fit may be unsuccessful; "
+                          "Maximum number of iterations reached.",
+                          AstropyUserWarning)
+        return res['x'], {info_name: res[res_name] for info_name, res_name in zip(['final_func_val', 'numiter', 'num_function_calls', 'exit_mode'],
+                                                                                  ['fun', 'nit', 'nfev', 'status']) if res_name in res}
```
Comment (nden) [inline astropy/modeling/optimizers.py]: Line too long

File: astropy/modeling/tests/test_fitters.py [hunk 15]
```diff
@@ -501,6 +501,21 @@ def test_param_cov(self):
...
         assert_allclose(olscov, fitter.fit_info['param_cov'])
 
+    def test_minimize(self):
+        """
+        Runs MinimizeFitter with most methods.
+        """
+        # get rid of tiny bound so the method which don't support
+        # bounds can be tested
+        self.gauss.bounds['stddev'] = (None, None)
+        for meth in ['Nelder-Mead', 'Powell', 'CG', 'BFGS', 'L-BFGS-B', 'TNC', 'SLSQP']:
```
Comment (nden) [inline astropy/modeling/tests/test_fitters.py]: It may be a bit better in terms of figuring out future regressions if this was written as a pytest parametrization.

```
methods = [...]
@pytest.mark.parametrize('method',  methods)
def test_minimize(self, method):
```

File: astropy/modeling/fitting.py [hunk 16]
```diff
@@ -1156,6 +1154,62 @@ def _convert_input(x, y, z=None, n_models=1, model_set_axis=0):
     return farg
 
 
+class MinimizeFitter(Fitter):
```
Comment (nden) [inline astropy/modeling/fitting.py]: Regarding the name of the class - all other fitter names follow the pattern of  `<OptimizerName><StatisticName>Fitter`.
This fitter uses `leastsquare` statistic with no option to change it. Following the naming convention should the name be `MinimizeLSQFitter`?

**Comment by nden** [comment]
@nocturnalastro This looks great. I left a few inline comments and change requests. 

This file also needs a PEP8 check but let's do it after the PR is approved and ready to merge so that it doesn't clutter the changes.

**Comment by astrofrog** [comment]
@nocturnalastro - it looks like there are still some real test failures, could you investigate? Thanks!

**Comment by bsipocz** [comment]
@nocturnalastro - this PR gone stale again. Could you please rebase and check the test failures that were mentioned above?
Thanks.

**Comment by nocturnalastro** [comment]
its seems that some of the failures are due to the tolerance parameter in minimize others (namely 'SLSQP' are producing a really poor fit probably due to a lack of constraints in the test) I will keep looking into it.
Sorry for the delay.
